Convert $.ajax() to fetch()

Geo1088 opened the pull request:
Yet another WIP PR from me, fixes #45.

With this change, the API for dealing with requests from modules will change a bit:

- `TBApi.sendRequest()` will resolve to a `Response` regardless of response error code, and will reject an `Error` on network error or request timeout/abortion. This mirrors `fetch()` semantics. Error objects will be always be instances of `Error` rather than e.g. `TypeError` or `NetworkError` because deserializing error objects from JSON is hard without using `eval()`.

- Other request methods will resolve to a `Response` only if the response status is OK. If a non-OK response is received, it will reject with an `Error` object, with the raw response attached as a `response` property. Other network errors that indicate no response was received will be rejected as normal `Error` objects with no `response` property, the same as with `TBApi.sendRequest`.

For example:
```js
TBApi.apiOAuthGET(...).then(async response => {
	// Action was successful!
}).catch(error => {
	if (error.response) {
		// We can check for error codes returned by Reddit here
		self.error('request failed with status', error.response.status);
	} else {
		// The request probably never even got to Reddit
		self.error('non-Reddit error on request:', error.message);
	}
});
```

Additionally, the `makeRequest` function will be made available in the background page. Its API is identical to `TBApi.sendRequest`.

---

A summary of structural changes is in [this comment](https://github.com/toolbox-team/reddit-moderator-toolbox/pull/250#issuecomment-597419446).

Known bugs:

- ~~Pulling HTML from JSON is broken somehow, resulting in stuff like the context popups not displaying things correctly. Not sure why this is.~~ Fixed with 6c90bab

Geo1088 commented on extension/data/background/handlers/webrequest.js:
```diff
@@ -21,104 +22,120 @@ async function getOAuthTokens (tries = 1) {
-            resolve({jqXHR, textStatus, errorThrown});
-        });
-    });
+function makeRequest (url, options) {
```
Commented above, will be used for ratelimit stuff later so might as well keep it in for now. Less work for a future PR.

creesch replied:
Gotcha 👍

Geo1088 commented on extension/data/tbapi.js:
```diff
@@ -11,26 +11,31 @@
+            body,
+            oauth,
+        });
+        if (response.error) {
```
Yes, this implementation is very WIP and nothing will work in its current state.

creesch replied:
Well you asked me to go over it :P Other than that I don't spot anything wrong with it at the moment.

Geo1088 commented:
At long last, ready for review! This is another big diff, so here's a summary to make reviewing easier.

- [`extension/data/background/handlers/webrequest.js`](https://github.com/toolbox-team/reddit-moderator-toolbox/pull/250/files#diff-0c26dbb2bf6a8eece45cab254fac2dc8)

  - `getOAuthTokens` was simplified to be more readable.
  - `serializeResponse` was created to serialize `Response` objects to JSON so they can be passed to the content script.
  - `makeHeaderObject` was removed, as it's been replaced by part of `serializeResponse`.
  - `makeRequest` was modified to just be a wrapper around `fetch`. This function will contain more stuff later  when we start tackling ratelimiting in #248.
  - The message handler now takes `body` rather than `data`. This prevents confusion between data in the URL parameters and data in the request body.
  - The message handler now messages back either a `Response` serialized to a JSON array, or a JSON object with an `error` key.
  - Various details of the actual requests were changed to match Fetch APIs.

- [`extension/data/tbapi.js`](https://github.com/toolbox-team/reddit-moderator-toolbox/pull/250/files#diff-67e55e5d4c2a111a861c21c89a373e0a)
  - `sendRequest` now takes `body` and `query` options rather than a `data` option for the same reason as above.
  - `sendRequest` has a new `okOnly` option. When set to true, the method will throw an error if a non-2xx response error code is encountered. The error will have a `response` property in this case. Otherwise, all responses will be returned regardless of error code, and the only errors thrown will be request interruptions/timeouts/abortions.
  - `apiOAuthRequest` was removed, as it was useless and only used to define `apiOAuthPOST` and `apiOAuthGET`.
  - `getJSON` and `post` were reimplemented with the above changes to lower-level request methods. They still return JSON objects, but now throw `Error` objects rather than the strange object stuff they threw before.
  - `apiOAuthGET` and `apiOAuthPOST` were also reimplemented. They no longer return strange custom objects, instead returning a `Response`. They also throw `Error`s rather than custom objects.

- [`extension/data/modules/historybutton.js](https://github.com/toolbox-team/reddit-moderator-toolbox/pull/250/files#diff-170550628f2608d5d6e5d7d99040a3bd)
  - Corrected error handling. When `.catch` precedes `.then` and the `.catch` handler doesn't re-throw, the `.then` is triggered even on error. To compensate for this, we return early if there is no data.

- [`extension/data/modules/modmatrix.js`](https://github.com/toolbox-team/reddit-moderator-toolbox/pull/250/files#diff-12edf01b814a98421bbdc86bf2053eb0)
- [`extension/data/modules/notifier.js`](https://github.com/toolbox-team/reddit-moderator-toolbox/pull/250/files#diff-42045f70544a35292ee6ab4fe4fe8025)
  - Updated response and error handling to work with the new expected values (`Response` objects and `Error` objects).

creesch commented:
Something somewhat related. Am I supposed/allowed to use `makeRequest()` if I want to do requests from a different handler in the background?

Geo1088 commented:
Hm, I didn't really consider that use case. I think I'm gonna change around how things are structured a bit to make that easier to deal with.

Geo1088 commented:
> Am I supposed/allowed to use makeRequest() if I want to do requests from a different handler in the background?

Latest commit, bdd68a2, addresses this. `makeRequest` now handles all options passed in the background page, which means yes, you can just use `makeRequest()` from anywhere else in the background page and pass it the same options you would pass to `TBApi.sendRequest()` in content scripts.

On the background page, `makeRequest` handles all options and makes the request, while the actual message listener adds response serialization for sending the response to the content script. In the content script, `TBApi.sendRequest()` is now solely responsible for sending the request options and deserializing the response.

creesch commented:
I assume this is still ready for review? Planning to look at this in the next couple of weeks so we can get started with the next release.

Also tagged @SpyTec as this is one of those things we probably want to have an extra set of eyes on.

SpyTec commented on extension/data/background/handlers/webrequest.js:
```diff
@@ -21,104 +22,143 @@ async function getOAuthTokens (tries = 1) {
-    let expired = false;
+
+    // Make sure the cookie is still valid
+    let validCookie = true;
```
Default to `false` and skip the else-statement

SpyTec commented on extension/data/background/handlers/webrequest.js:
```diff
@@ -21,104 +22,143 @@ async function getOAuthTokens (tries = 1) {
+        await makeRequest('https://mod.reddit.com/mail/all');
+        return getOAuthTokens(tries + 1);
+    } else {
+        // If we tried that 3 times and still no dice, the user probably isn't logged
```
Can't we simply check response code for `401` instead of fetching modmail 3 times?

creesch replied:
No we do this because reddit can be unstable at times and the code we can get in return isn't always reliable.

SpyTec replied:
Typical :P

SpyTec commented on extension/data/background/handlers/webrequest.js:
```diff
@@ -21,104 +22,143 @@ async function getOAuthTokens (tries = 1) {
+messageHandlers.set('tb-request', requestOptions => makeRequest(requestOptions)
+    .then(async response => {
+        response = await serializeResponse(response);
+        return {response};
```
Do we need await here? And there's no need to re-assign response, `serializeResponse` can just be returned

Geo1088 replied:
Await is required because it's being returned as an object with a property. Goal of the reassignment was to increase readability somewhat, but it's a weird structure that's hard to represent clearly because of terminology overlap. Equivalent alternative would be `return {response: await serializeResponse(response)}`, take your pick.

SpyTec replied:
I definitely prefer the `{response: await serializeResponse(response)}`. Re-assigning the param is bad practice

SpyTec commented on extension/data/modules/historybutton.js:
```diff
@@ -393,6 +393,9 @@ function historybutton () {
             $error.html('unable to load userdata</br>shadowbanned?');
             TB.ui.longLoadNonPersistent(false);
         }).then(d => {
```
Unrelated to PR, is `d` for `data`? Shouldn't have to guess this

Geo1088 replied:
Probably, but won't pretend to know for sure, lol

SpyTec commented on extension/data/modules/modmatrix.js:
```diff
@@ -434,9 +434,9 @@ function modmatrix () {
-                    self.log(`Mod log request ${requestData.count}to ${requestData.count + requestData.limit} failed (${jqXHR.status}), ${textStatus}: ${errorThrown}`);
-                    if (jqXHR.status === 504) {
+                .catch(error => {
+                    self.log(`Mod log request ${requestData.count}to ${requestData.count + requestData.limit} failed:`, error);
```
```suggestion
                    self.log(`Mod log request ${requestData.count} to ${requestData.count + requestData.limit} failed:`, error);
```

SpyTec commented on extension/data/tbapi.js:
```diff
@@ -11,88 +11,105 @@
      * @param {object} options The options for the AJAX request
-     * @param {string} options.method The HTTP method to use for the request
+     * @param {string} [options.method] The HTTP method to use for the request
      * @param {string} options.endpoint The endpoint to request
```
Switch `method` and `endpoint` around to have all optional params grouped

SpyTec commented on extension/data/tbapi.js:
```diff
@@ -11,88 +11,105 @@
-            throw response;
+    TBApi.sendRequest = async ({method, endpoint, query, body, oauth, okOnly}) => {
+        // Make the request
+        const messageReply = await browser.runtime.sendMessage({
```
Both old and new code never catches in the even it returns an error. [MDN says](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/runtime/sendMessage):

> If an error occurs while connecting to the extension, the promise will be rejected with an error message

Should we not catch this?

Geo1088 replied:
No, the consumer of sendRequest is meant to handle that case. No need to explicitly re-throw.

SpyTec commented on extension/data/tbcore.js:
```diff
@@ -1812,61 +1812,43 @@ function initwrapper ({userDetails, newModSubs, cacheDetails}) {
+        TBApi.getJSON('/subreddits/mine/moderator.json', {
+            after,
+            limit: 100,
+        }).then(data => {
```
Shouldn't this be `response`? Accessing `data.data` is confusing

Geo1088 replied:
`response.data` is potentially confusing because `getJSON` doesn't return a `Response`, but the JSON body. Maybe `json.data` would be clearer?

creesch commented:
I am getting some errors, but I think those are the result of something already resolved in master related to promisfying some stuff. Can you merge it in so we can double check? 

![image](https://user-images.githubusercontent.com/2502190/77849079-49475c00-71c9-11ea-993c-1f90754bc750.png)

Geo1088 commented:
Done @creesch :+1:

Geo1088 commented:
Context popup issues were caused by bad querystring handling, fixed with latest commit. Can't identify any other bugs with this branch, let's get some final 👀 on this and it should be good to merge!

creesch commented:
Just did a testdrive hitting a bunch of API related functionality and it works all fine now. I'll await the changes you mentioned that spytec requested and after those (and a tiny bit of testing) I am sure we can merge this.
